WebGLRenderer: Auto shader output definitions for MRT

vanruesc opened the pull request:
### Description

This PR aims to make Multiple Render Targets (MRT) easier to use by implementing automatic shader output definitions for built-in materials and `ShaderMaterial`.

```glsl
// Example output definitions:
layout(location = 0) out lowp vec4 out_FragData0;
#define out_Color out_FragData0
layout(location = 1) out mediump vec4 out_FragData1;
#define out_Normal out_FragData1
```

#### Context

For MRT workflows it's up to the user to define what data should be written to the respective textures. The [MRT example](https://threejs.org/examples/?q=mrt#webgl2_multiple_rendertargets) promotes the use of `RawShaderMaterial` to achieve this. However, MRT is most useful for rendering out additional pixel data during the main render pass. In most situations the main scene will consist of meshes that use built-in materials like `MeshStandardMaterial`, but those materials are currently difficult to use with MRT. 

#### Improving compatibility

It's possible to use `onBeforeCompile` on every object in a scene to modify built-in materials, but this has some downsides:

- An object's material may already have an `onBeforeCompile` hook, so care must be taken to preserve it.
- All objects need to be tracked to avoid assigning `onBeforeCompile` hooks multiple times.
- Objects and materials may be added or removed at runtime which requires additional scene processing to ensure that all materials have the required `onBeforeCompile` hooks.

If it wasn't necessary to define shader outputs, we could avoid using `onBeforeCompile` altogether and modify the `ShaderChunk` library instead ([example](https://github.com/pmndrs/postprocessing/blob/a957328a92beae5a92982dbe653e8ceb82aa592a/src/shader-chunks/ShaderChunkExtensions.ts#L61-L65)). But at the moment there's [no way around](https://github.com/pmndrs/postprocessing/blob/a957328a92beae5a92982dbe653e8ceb82aa592a/src/passes/GeometryPass.ts#L332-L348) `onBeforeCompile`.

#### Meeting halfway

With the changes in this PR, three can define shader outputs automatically based on the current render target. To avoid breakage, this is only done for built-in materials and for `ShaderMaterial` if the shader doesn't already define shader outputs. This implementation also uses the names of the MRT textures to define macros of the form `out_${name}` for convenience which allows for conditional MRT code like the following: 

```glsl
#ifdef out_Normal

	out_Normal = vec4(normal, 1.0);

#endif
```

I've considered using the texture names as-is but opted for a prefix approach to reduce the risk of name conflicts.

#### Open issue

~~This PR is in draft mode because the material/program doesn't get updated after switching the render target. I've tried modifying WebGLPrograms.getProgramCacheKey to account for the render target, but that didn't work out. Setting material.needsUpdate to true also didn't trigger a program update.~~

vanruesc committed:
Generate shader outputs

vanruesc committed:
Update webgl2_multiple_rendertargets.html

vanruesc commented on examples/webgl2_multiple_rendertargets.html:
```diff
@@ -82,7 +40,7 @@
 				return vec4( mix( pow( value.rgb, vec3( 0.41666 ) ) * 1.055 - vec3( 0.055 ), value.rgb * 12.92, vec3( lessThanEqual( value.rgb, vec3( 0.0031308 ) ) ) ), value.a );
 			}
 
-			layout(location = 0) out vec4 pc_FragColor;
```
Btw: does anyone know what "pc" stands for? :thinking:

yfunk replied:
My best guess would be something in the direction of "program compatibility", since `pc_fragColor` is related to GLSL1 compatibility in `WebGLProgram`.

vanruesc replied:
Oh yeah, that could be it :+1: 

I just found it weird that I couldn't find anything when I tried looking it up.

vanruesc committed:
Check for out instead of layout

vanruesc committed:
Adjust outputDefinition RegExp

vanruesc committed:
Update webgl2_rendertarget_texture2darray.html

vanruesc committed:
Update webgl2_multiple_rendertargets.jpg

vanruesc committed:
Improve RegExp

puxiao commented:
> **Some checks were not successful**
>
> DeepScan：1 new and 1 fixed issues
>
> Local variable 'xr' is not used.
>
> editor/js/Viewport.js
>
>  ....


@vanruesc  This issue has been fixed, you need to pull the code again.

<br>



**Solution:**

```
git checkout auto-shader-outputs
git remote add upstream https://github.com/threejs/three.js.git
git fetch upstream
git merge upstream/auto-shader-outputs
```



... All checks have passed ...

vanruesc committed:
Merge branch 'dev' into auto-shader-outputs

vanruesc committed:
Merge branch 'dev' into auto-shader-outputs

vanruesc committed:
Merge branch 'dev' into auto-shader-outputs

vanruesc committed:
Organize imports

vanruesc committed:
Include render target in program cache key

vanruesc committed:
Clean up

vanruesc committed:
Update webgl_multiple_rendertargets.html

vanruesc committed:
Remove render target from program parameters

vanruesc committed:
Revert "Remove render target from program parameters"

This reverts commit c3eb0cf370b4b8b926c04474697d3b3bbc0295f0.

vanruesc commented:
> This PR is in draft mode because the material/program doesn't get updated after switching the render target. I've tried modifying `WebGLPrograms.getProgramCacheKey` to account for the render target, but that didn't work out. Setting `material.needsUpdate` to `true` also didn't trigger a program update.

I've fixed this now by including the render target in the material properties to detect if the current render target differs from the one that was previously associated with the material. The `needsUpdate` flag did in fact work before, but I didn't realize that I had to call `render()` manually in the demo I was testing with.

I've adjusted `webgl_multiple_rendertargets` to test switching between render targets and it appears to be working as expected. One downside is that render targets are only checked superficially. If the texture attachments are changed, materials will not be updated unless `needsUpdate` is set.

vanruesc committed:
Formatting

vanruesc committed:
Merge branch 'dev' into auto-shader-outputs

vanruesc committed:
Fix shader outputs

vanruesc committed:
Remove unnecessary swizzles

vanruesc committed:
Merge branch 'dev' into auto-shader-outputs

vanruesc committed:
Fix VSM depth write

vanruesc committed:
Only set red channel for basic depth

Looks like VSM uses RG and PCF RGBA

vanruesc committed:
Add comment

vanruesc committed:
Update webgl_shadow_contact.html

vanruesc committed:
Merge branch 'dev' into auto-shader-outputs

vanruesc committed:
Minimize friction

vanruesc committed:
Update doc comment

vanruesc committed:
Merge branch 'dev' into auto-shader-outputs
